### PR TITLE
Try local signature verification first before downloading

### PR DESCRIPTION
**Pull request by phmccarty**
In case the Manifest.MoM.sig already exists locally (e.g. it was shipped
inside the Manifest.MoM.tar), then try to verify the local copy first.
If the verify fails because the .sig is missing, or for any other
reason, then download a fresh copy from the server.

For the implementation, I refactored verify_signature() to not print
error messages if the verification fails the first time, because we
"retry" by downloading the fresh copy. If the verification fails the
second time, then error messages are printed. Also, I consolidated the
cleanup logic in this function to reduce duplication.

Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>

### CODE REVIEW

File: src/signature.c [hunk 1]
```diff
@@ -390,10 +377,11 @@ int verify_callback(int ok, X509_STORE_CTX *stor)
 	return ok;
 }
 
-/* Downloads the corresponding signature filename from the
- * swupd server.
+/* Verifies signature for the file DATA_FILENAME, or downloads the signature
+ * first from DATA_URL and verifies the signature afterward.
```
Comment (tmarcu) [inline src/signature.c]: I think the comment on download would be better stated saying: "Verifies signature for the local file DATA_FILENAME first, and on failure downloads the signature based on DATA_URL and tries to verify again."

Comment (phmccarty) [inline src/signature.c]: Sounds good, will change that

**Comment by phmccarty** [comment]
Updated patch set 2

File: src/signature.c [hunk 2]
```diff
@@ -118,145 +118,132 @@ void terminate_signature(void)
...
-	BIO *sig_BIO;
+	BIO *sig_BIO = NULL;
 
-	PKCS7 *p7;
-	BIO *verify_BIO;
+	PKCS7 *p7 = NULL;
+	BIO *verify_BIO = NULL;
 
 	/* get the signature */
 	sig_fd = open(sig_filename, O_RDONLY);
```
Comment (bradTpeters) [inline src/signature.c]: Why open() before fstat()'ing the file? The opposite order would make more sense IMO

**Comment by bradTpeters** [comment]
New flow and retry looks good.  +1

File: src/signature.c (hunk 2 above)
Comment (phmccarty) [inline src/signature.c]: The fstat() is done to get the file size actually... And calling open() first avoids TOCTOU issues.

Comment (bradTpeters) [inline src/signature.c]: Ah, thanks for clarifying

**Comment by phmccarty** [comment]
Cool, will merge shortly

**Comment by phmccarty** [comment]
Rebased and merged

**Pull request closed without merging**